aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp

```diff
@@ -197,3 +205,76 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
 } // namespace at
 
 #endif // USE_FBGEMM
+
+
+torch::jit::class_<LinearPackedParamsBase> register_linear_params() {
+  using SerializationType = std::tuple<at::Tensor, c10::optional<at::Tensor>, std::string, std::string>;
+  static int custom_class_handler = torch::jit::register_custom_class_handler();
+  static auto register_linear_params =
+      torch::jit::class_<LinearPackedParamsBase>("LinearPackedParamsBase")
+          .def_pickle(
+              [](const c10::intrusive_ptr<LinearPackedParamsBase>& params)
+                  -> SerializationType { // __getstate__
+                at::Tensor weight;
+                c10::optional<at::Tensor> bias;
+                std::tie(weight, bias) = params->unpack();
+                return std::make_tuple(
+                    std::move(weight),
+                    std::move(bias),
+                    params->backend(),
+                    params->bit_width());
+              },
+              [](SerializationType state)
+                  -> c10::intrusive_ptr<
+                      LinearPackedParamsBase> { // __setstate__
+                at::Tensor weight;
+                c10::optional<at::Tensor> bias;
+                std::string backend, bit_width;
+                weight = std::move(std::get<0>(state));
+                bias = std::move(std::get<1>(state));
+                backend = std::move(std::get<2>(state));
+                bit_width = std::move(std::get<3>(state));
+
+#ifdef USE_FBGEMM
+                if (backend == "FBGEMM") {
```

Comment:
we shouldn't serialize the backend and instead use qengine do deside which one to call. You probably don't need branching here and instead you can just invoke the linear_prepack implementation directly, it already does it: https://github.com/pytorch/pytorch/blob/f61b45fc89086d773d099112578bdbc3db13da78/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp#L180

Reply:
I don't think a global qengine is the correct design overall, I've already had to come in and debug issues that that's introduced, and it just seems like it was not a well-thought-out design decision